### DISCUSSION TITLE
make streamlines on a 2D mesh

### DISCUSSION THREAD

**Discussion by emi0000** [pyvista/pyvista#5227] [Q&A]
Hi,

I'm trying to plot streamlines on a 2D mesh. The vector field does not have a z component, so the lines should be restricted to the plane.

      import numpy as np
      import pyvista as pv

      plane=pv.Plane().triangulate()
      plane.point_data['vectors']=np.array([-plane.points[:,1],plane.points[:,0],0*plane.points[:,2]]).T
      plane.points[:,2]=0*plane.points[:,2]
      plane.set_active_vectors('vectors')

      seed=pv.Line(pointa=(0., 0.1, 0.0), pointb=(0., 0.4, 0.0), resolution=10)
      stream = plane.streamlines_from_source(seed,vectors='vectors', max_time=100, initial_step_length=0.01, integration_direction="both",surface_streamlines=True)
     
      pl=pv.Plotter()
      pl.add_mesh(plane,show_edges=True,color='grey')
      pl.add_arrows(plane.points,plane.point_data['vectors'],color='yellow',mag=0.3 )
      pl.add_mesh(stream.tube(radius=0.1),color='red')
      pl.show()

The mesh and the vector field looks as expected, but there seems to be no streamline (empty dataset).
Any ideas whats going wrong?

**Comment by MatthewFlamm** [comment]
Your code generates valid streamlines for me.  Is the `stream` object empty for you?  Can you share `pv.Report()`?

**Reply by emi0000** [reply]
thanks

![pvreport](https://github.com/pyvista/pyvista/assets/88888989/3a4fd178-f095-44e2-affa-28ab8cef2e07)


I'm getting:

2023-11-26 20:31:44.268 (6216.949s) [        F62F7000]    vtkExtractEdges.cxx:435   INFO| Executing edge extractor: points are renumbered
2023-11-26 20:31:44.268 (6216.949s) [        F62F7000]    vtkExtractEdges.cxx:551   INFO| Created 320 edges
/usr/local/lib/python3.10/dist-packages/pyvista/jupyter/pv_pythreejs.py:519: UserWarning: Empty or unsupported dataset <class 'pyvista.core.pointset.PolyData'>.
  warnings.warn(f'Empty or unsupported dataset {type(dataset)}.')

print(stream) gives:

PolyData (0x7ff6bf2aefe0)
  N Cells:	202
  N Points:	186204
  N Strips:	0
  X Bounds:	-4.000e-01, 4.000e-01
  Y Bounds:	-4.000e-01, 4.000e-01
  Z Bounds:	0.000e+00, 0.000e+00
  N Arrays:	9

So it seems that the plot makes the problems. I'm using a jupyter lab (3.5.3) in a docker of dolfinx. I was not able to make it work with other backends (e.g. pv.set_jupyter_backend('ipygany') gives an error message: only size-1 arrays can be converted to Python scalars)

it works with
pl.add_points(stream.points,color='red')

How can I access all points of a cell? 
Or how can I access the different stremlines?
e.g. seed=pv.Line(pointa=(0., 0.1, 0.0), pointb=(0., 0.4, 0.0), resolution=3) gives 4 different streamlines but 8 cells

**Reply by MatthewFlamm** [reply]
You are using an old version on PyVista.  You could try upgrading, but can you also share what the plot looks like?  Your code generates streamlines, but they overlap so significantly that they don't look like typical streamlines when I plot.

**Reply by emi0000** [reply]
If you use 

      seed=pv.Line(pointa=(0., 0.1, 0.0), pointb=(0., 0.4, 0.0), resolution=3) 
      stream = plane.streamlines_from_source(seed,vectors='vectors', max_time=100, initial_step_length=0.01,surface_streamlines=True)

and later in the plot
      
      pl.add_points(stream.points,color='red')

you will get 4 different stream lines and they are well separated and fit the vector plot well, so they are correct.

Originally I increased the resolution in the seed command, as I was wondering whether the start points might be the problem. 

If I try to plot with

      pl.add_mesh(stream.tube(radius=0.1),color='red') 

![Figure_1](https://github.com/pyvista/pyvista/assets/88888989/69d0be3e-b22e-46b8-be41-d4ce32852322)


Plotting with  

pl.add_mesh(stream.tube(radius=0.1),color='red')

fails (vector field is correct and mesh is correct but no streamlines), so there are no streamlines at all, but the message

![Untitled](https://github.com/pyvista/pyvista/assets/88888989/ac5218a9-a049-4af6-b49f-f85dd203b4de)


pv_pythreejs.py:519: UserWarning: Empty or unsupported dataset <class 'pyvista.core.pointset.PolyData'>.
warnings.warn(f'Empty or unsupported dataset {type(dataset)}.')

How are the streamlines organized in the PolyData class. As I get 8 cells for the case of 4 streamlines I was wondering how to access the single streamlines computed from the starting point up to max_time? So I was expecting somehow to find 4 "Elements".

**Reply by MatthewFlamm** [reply]
I suspect this is a `pythreejs` problem rather than a PyVista problem, particularly since I have no problem rendering in PyVista alone.  There must be some issue with highly overlapped meshes in `pythreejs`.  Can you either render directly in `pyvista` (presumably by not using Jupyter notebooks) or use the `trame` backend?

> How are the streamlines organized in the PolyData class. As I get 8 cells for the case of 4 streamlines I was wondering how to access the single streamlines computed from the starting point up to max_time? So I was expecting somehow to find 4 "Elements".

What you are seeing is that you are using `integration_direction="both"` (note that this is the default).  There are 2 cells for each starting point, one that goes forwards and one that goes backwards.  You can get the individual streamlines by using `get_cell`.  For example:

```py
stream.get_cell(0)
stream.get_cell(0).points
```

Forward looking enhancements to pyvista that could make this part easier:  Unfortunately, a PyVista `Cell` does not today allow to casting it to a `PolyData` object which you would need to make it a tube.  The other possibility is that the `tube` filter should also be available on an `UnstructuredGrid` as there is a `cast_to_unstructured_grid` method for the cell, i.e. `stream.get_cell(0).cast_to_unstructured_grid()`.

**Reply by MatthewFlamm** [reply]
Of course the other option to get a single cell per starting point is to run `streamlines_from_source` individually for each point (and possibly going forwards and then backwards).

**Reply by MatthewFlamm** [reply]
>  The other possibility is that the tube filter should also be available on an UnstructuredGrid as there is a cast_to_unstructured_grid method for the cell, i.e. stream.get_cell(0).cast_to_unstructured_grid().

This isn't possible.  We should look into whether we can add a `cast_to_polydata` method for a cell that is 0D, 1D, or 2D.

Edit: tracked in #5238

**Reply by emi0000** [reply]
I tried to install trame, but within jupiter lab I get the error

Invalid Jupyter notebook plotting backend "trame"

As I'm working in a docker, I'm not sure how to start the code without jupyter lab. Sure I can open a terminal in the docker container and start the code, but that will not open a plot window, it just crashes.  Do you know how I can directly plot to a file with pyvista, without opening a plot window. 

The command

      stream.get_cell(0)

ends up in a error  'PolyData' object has no attribute 'get_cell'

With the option

      integration_direction='forward'   

in streamlines_from_source I'm getting now 4 cells for the 4 streamlines above.

**Reply by MatthewFlamm** [reply]
`get_cell` is a newer feature, you will have to upgrade PyVista.  This is likely also the problem with `trame` support.

To plot completely off screen to a file, I do something like this:

```py
import pyvista as pv
pl = pv.Plotter(off_screen=True)
pl.add_mesh(pv.Sphere())
pl.show(screenshot="sphere.png")
```

**Reply by emi0000** [reply]
I now have a newer pyvista version

pyvista 0.42.2
vtk 9.3.0

also jupyter lab is more recently, version 4.08, as I took the latest stable docker image of dolfinx/lab.

I had a few problemes but most of them were solved by installing

trame 
trame-vuetify
trame-vtk

However I now get either that the kernel died or that the connection is refused or that the requested page can not be found (error 404). So probably thats a problem of opening a window from within docker. There are some hints on the pyviste dokumentation, but I was not able to make it work.

I set also

pv.global_theme.jupyter_backend trame
pv.global_theme.trame.server_proxy_prefix /proxy/
pv.global_theme.trame.server_proxy_enabled True

and started pv.start_xvfb(), but it remains the same problem.

**Reply by MatthewFlamm** [reply]
This is now out of my experience and hopefully someone else can jump in.

**Reply by emi0000** [reply]
thanks again, do you think I have to make a new topic?

**Reply by MatthewFlamm** [reply]
Up to you, but the context here might still be useful

**Reply by emi0000** [reply]
see #5261
